File: ScreenControl_Unity/Assets/ScreenControl/Client/Scripts/Cursors/DotCursor.cs
using System.Collections;
using UnityEngine;
using Ultraleap.ScreenControl.Client.Connection;

namespace Ultraleap.ScreenControl.Client.Cursors
{
    // Class: DotCursor
    // This is an example Touchless Cursor which positions a dot on the screen at the hand location,
    // and reacts to the current ProgressToClick of the action (what determines this depends on the
    //  currently active interaction).
    //
    // We recommend using the "RingCursor" Prefab which is set up for use with this class.
    public class DotCursor : TouchlessCursor
    {
        // Group: Variables

        // Variable: fadeDuration
        // The number of frames over which the cursor should fade when appearing/disappearing
        [Range(0f, 60f)] public float fadeDuration = 30;

        // Variable: cursorDotSize
        // The size of the dot when it isn't being shrunk
        [SerializeField]
        protected float cursorDotSize = 0.25f;

        // Variable: cursorBorder
        // The image of the border around the dot, this is the parent image in the prefab and is
        //  used to do all of the scaling of the images that make up this cursor.
        [Header("Graphics")]
        public UnityEngine.UI.Image cursorBorder;

        // Variable: cursorFill
        // The main background image of the Cursor, used for fading the image out.
        public UnityEngine.UI.Image cursorFill;

        // Variable: ringOuterSprite
        // This refers to the Ring around the central cursor that is used to display the "reactive"
        // state of the cursor; the closer the ring is to the dot, the closer you are to "clicking".
        //
        // The sprite is accessed to fade in/out the ring as it is needed.
        public SpriteRenderer ringOuterSprite;

        // Variable: ringEnabled
        // Enables/disables the ring cursor around the dot. Here primarily for use in the inspector.
        [Header("Ring")]
        public bool ringEnabled;

        // Variable: cursorMaxRingSize
        // The maximum size for the ring to be relative to the size of the dot.
        //
        //  e.g. a value of 2 means the ring can be (at largest) twice the scale of the dot.
        public float cursorMaxRingSize = 8;

        // Variable: ringCurve
        // This curve is used to determine how the ring's scale changes with the value of the latest
        // InputAction's ProgressToClick.
        public AnimationCurve ringCurve;

        // Variable: ringOuter
        // This refers to the Ring around the central cursor that is used to display the "reactive"
        // state of the cursor; the closer the ring is to the dot, the closer you are to "clicking"
        public RectTransform ringOuter;

        // Variable: ringMask
        // This is a reference to the mask that is used to make the center of the ring visual
        // transparent. It has to be scaled to match the ring itself.
        public RectTransform ringMask;

        // Variable: ringThickness
        // Used to set the thickness of the ring itself (i.e. the distance between the inner and
        // outer edges of the ring)
        public float ringThickness = 2;

        // Variable: pulseShrinkCurve
        // When a "click" is recognised, an animation plays where the dot "pulses" (briefly
        // shrinking and expanding). This AnimationCurve governs the shrinking that follows a
        // "DOWN" input.
        [Header("Pulse")]
        public AnimationCurve pulseShrinkCurve;

        // Variable: pulseGrowCurve
        // When a "click" is recognised, an animation plays where the dot "pulses" (briefly
        // shrinking and expanding). This AnimationCurve governs the growing back to full size
        // that follows a "UP" input.
        public AnimationCurve pulseGrowCurve;

        // Variable: pulseSeconds
        // The amount of time each part of the pulse should last (a pulse will last twice this long)
        [Range(0.01f, 1f)] public float pulseSeconds;

        // Variable: cursorDownScale
        // The scale of the cursor's dot at its smallest size (between the shrink and the grow )
        [Range(0.01f, 2f)] public float cursorDownScale;

        Coroutine cursorScalingRoutine;

        protected float maxRingScale;
        public Color dotFillColor;
        public Color dotBorderColor;
        public Color ringColor;

        protected bool hidingCursor = true;
        protected bool growQueued = false;
        protected Vector3 cursorLocalScale = Vector3.one;
        protected const int ringSpriteSortingOrder = 32766;

        // Group: Functions

        // Function: UpdateCursor
        // Used to update the cursor when recieving a "MOVE" <ClientInputAction>. Updates the
        // cursor's position, as well as the size of the ring based on the current ProgressToClick.
        public void UpdateCursor(Vector2 _screenPos, float _progressToClick)
        {
            targetPos = _screenPos;

            if (ringEnabled)
            {
                if (hidingCursor)
                {
                    ringOuterSprite.color = new Color(ringColor.r, ringColor.g, ringColor.b, 0);
                }
                else
                {
                    //progressToClick is between 0 and 1. Click triggered at progressToClick = 1
                    _progressToClick = Mathf.Clamp01(1f - _progressToClick);

                    // 0.9f so that the boundary between ring and dot is not visible - small overlap.
                    float minRingScale = 0.9f + ringThickness;
                    float ringScale = Mathf.Lerp(minRingScale, maxRingScale, ringCurve.Evaluate(_progressToClick));
                    ringOuterSprite.color = new Color(ringColor.r, ringColor.g, ringColor.b, Mathf.Lerp(1f, 0f, _progressToClick));

                    ringOuter.transform.localScale = Vector3.one * ringScale;

                    ringMask.transform.localScale = new Vector3()
                    {
                        x = Mathf.Max(0, ringOuter.localScale.x - ringThickness),
                        y = Mathf.Max(0, ringOuter.localScale.y - ringThickness),
                        z = ringOuter.localScale.z
                    };
                }
            }
        }

        // Group: TouchlessCursor Overrides

        // Function: HandleInputAction
        // This override replaces the basic functionality of the <TouchlessCursor>, making the
        // cursor's ring scale dynamically with the current ProgressToClick and creating a small
        // "shrink" animation when a "DOWN" event is recieved, and a "grow" animation when an "UP"
        // is recieved.
        protected override void HandleInputAction(ClientInputAction _inputData)
        {
            switch (_inputData.InputType)
            {
                case InputType.MOVE:
                    UpdateCursor(_inputData.CursorPosition, _inputData.ProgressToClick);
                    break;

                case InputType.DOWN:
                    growQueued = false;
                    if (cursorScalingRoutine != null)
                    {
                        StopCoroutine(cursorScalingRoutine);
                    }
                    cursorScalingRoutine = StartCoroutine(ShrinkCursorDot());
                    break;

                case InputType.UP:
                    growQueued = true;
                    break;

                case InputType.CANCEL:
                    break;
            }
        }

        // Function: Update
        // This override runs the basic functionality of <TouchlessCursor> and also ensures that if
        // the cursor has a <growQueued> and has the ability to, it should start the "grow"
        // animation.
        protected override void Update()
        {
            base.Update();

            if (growQueued && cursorScalingRoutine == null)
            {
                growQueued = false;
                cursorScalingRoutine = StartCoroutine(GrowCursorDot());
            }
        }

        // Function: InitialiseCursor
        // This override ensures that the DotCursor is properly set up with relative scales and
        // sorting orders for the ring sprites.
        protected override void InitialiseCursor()
        {
            ConnectionManager.HandFound += ShowCursor;
            ConnectionManager.HandsLost += HideCursor;

            bool dotSizeIsZero = Mathf.Approximately(cursorDotSize, 0f);
            cursorDotSize = dotSizeIsZero ? 1f : cursorDotSize;
            cursorBorder.transform.localScale = new Vector3(cursorDotSize, cursorDotSize, cursorDotSize);
            SetCursorLocalScale(cursorDotSize);

            if (ringEnabled)
            {
                maxRingScale = (1f / cursorDotSize) * cursorMaxRingSize;

                // This is a crude way of forcing the sprites to draw on top of the UI, without masking it.
                ringOuterSprite.sortingOrder = ringSpriteSortingOrder;
                ringMask.GetComponent<SpriteMask>().isCustomRangeActive = true;
                ringMask.GetComponent<SpriteMask>().frontSortingOrder = ringSpriteSortingOrder + 1;
                ringMask.GetComponent<SpriteMask>().backSortingOrder = ringSpriteSortingOrder - 1;
            }
        }

        // Function: ShowCursor
        // This override replaces the basic functionality of the <TouchlessCursor> to ensure that the cursor is faded smoothly
        // when being set to show.
        public override void ShowCursor()
        {
            bool wasShowing = !hidingCursor;
            hidingCursor = false;

            if (wasShowing)
            {
                return;
            }

            ResetCursor();
            StartCoroutine(FadeCursor(0, 1, fadeDuration));
            cursorBorder.enabled = true;
            cursorFill.enabled = true;

            if (ringEnabled)
            {
                ringOuterSprite.enabled = true;
                ringOuterSprite.color = new Color(ringColor.r, ringColor.g, ringColor.b, 0);
            }
        }

        // Function: HideCursor
        // This override replaces the basic functionality of the <TouchlessCursor> to ensure that the cursor is faded smoothly
        // when being set to hide.
        public override void HideCursor()
        {
            bool wasHiding = hidingCursor;
            hidingCursor = true;

            if (wasHiding)
            {
                return;
            }

            ResetCursor();
            StartCoroutine(FadeCursor(1, 0, fadeDuration, true));

            if (ringEnabled)
            {
                ringOuterSprite.color = new Color(ringColor.r, ringColor.g, ringColor.b, 0);
            }
        }

        // Group: Coroutine Functions

        // Function: GrowCursorDot
        // This coroutine smoothly expands the cursor dots size.
        public IEnumerator GrowCursorDot()
        {
            SetCursorLocalScale(cursorDownScale * cursorDotSize);

            YieldInstruction yieldInstruction = new YieldInstruction();
            float elapsedTime = 0.0f;

            while (elapsedTime < pulseSeconds)
            {
                yield return yieldInstruction;
                elapsedTime += Time.deltaTime;

                float scale = Utilities.MapRangeToRange(pulseGrowCurve.Evaluate(elapsedTime / pulseSeconds), 0, 1, cursorDownScale * cursorDotSize, cursorDotSize);
                SetCursorLocalScale(scale);
            }

            SetCursorLocalScale(cursorDotSize);
            cursorScalingRoutine = null;
        }

        // Function: ShrinkCursorDot
        // This coroutine smoothly contracts the cursor dots size.
        public IEnumerator ShrinkCursorDot()
        {
            YieldInstruction yieldInstruction = new YieldInstruction();
            float elapsedTime = 0.0f;

            while (elapsedTime < pulseSeconds)
            {
                yield return yieldInstruction;
                elapsedTime += Time.deltaTime;

                float scale = Utilities.MapRangeToRange(pulseShrinkCurve.Evaluate(elapsedTime / pulseSeconds), 0, 1, cursorDownScale * cursorDotSize, cursorDotSize);
                SetCursorLocalScale(scale);
            }

            SetCursorLocalScale(cursorDownScale * cursorDotSize);
            cursorScalingRoutine = null;
        }

        // Function: FadeCursor
        // This coroutine smoothly fades the cursors colours.
        private IEnumerator FadeCursor(float _from, float _to, float _duration, bool _disableOnEnd = false)
        {
            for (int i = 0; i < _duration; i++)
            {
                yield return null;
                float a = Mathf.Lerp(_from, _to, i / _duration);

                cursorBorder.color = new Color()
                {
                    r = cursorBorder.color.r,
                    g = cursorBorder.color.g,
                    b = cursorBorder.color.b,
                    a = a * dotBorderColor.a
                };

                cursorFill.color = new Color()
                {
                    r = cursorFill.color.r,
                    g = cursorFill.color.g,
                    b = cursorFill.color.b,
                    a = a * dotFillColor.a
                };
            };

            if (_disableOnEnd)
            {
                SetCursorLocalScale(cursorDotSize);

                cursorBorder.enabled = false;
                cursorFill.enabled = false;

                if (ringEnabled)
                {
                    ringOuterSprite.enabled = false;
                }
            }
        }

        // Function: SetCursorLocalScale
        // This function resizes the cursor and its border.
        private void SetCursorLocalScale(float _scale)
        {
            cursorLocalScale = new Vector3(_scale, _scale, _scale);
            cursorBorder.transform.localScale = cursorLocalScale;
        }

        // Function: ResetCursor
        // This function stops all scaling coroutines and clears their related variables.
        public void ResetCursor()
        {
            StopAllCoroutines();
            cursorScalingRoutine = null;
            growQueued = false;

            SetCursorLocalScale(cursorDotSize);
        }
    }
}

File: TF_Tooling_Web/src/Cursors/DotCursor.ts
import TouchFree from '../TouchFree';
import { TouchFreeInputAction, InputType } from '../TouchFreeToolingTypes';
import { MapRangeToRange } from '../Utilities';
import { TouchlessCursor } from './TouchlessCursor';

// Class: DotCursor
// This is an example Touchless Cursor which positions a dot on the screen at the hand location,
// and reacts to the current ProgressToClick of the action (what determines this depends on the
// currently active interaction).
export class DotCursor extends TouchlessCursor {
    // Set the update rate of the animation to 30fps.
    readonly animationUpdateDuration: number = (1 / 30) * 1000;

    // Group: Variables

    // Variable: cursorRing
    // The HTMLElement that visually represents the cursors ring.
    cursorRing: HTMLElement;

    // Variable: ringSizeMultiplier
    // The maximum size for the ring to be relative to the size of the dot.
    //
    // e.g. a value of 2 means the ring can be (at largest) twice the scale of the dot.
    ringSizeMultiplier: number;

    private cursorStartSize: Array<number>;
    private animationSpeed: Array<number> = [0, 0];

    private currentAnimationInterval = -1;

    private growQueued = false;

    private currentFadingInterval = -1;

    private dotCursorElement: HTMLElement;

    // Group: Functions

    // Function: constructor
    // Constructs a new cursor consisting of a central cursor and a ring.
    // Optionally provide an animationDuration to change the time it takes for the 'squeeze'
    // confirmation animation to be performed. Optionally provide a ringSizeMultiplier to change
    // the size that the <cursorRing> is relative to the cursor.
    //
    // If you intend to make use of the <WebInputController>, make sure that both cursor and
    // cursorRing have the "touchfree-cursor" class. This prevents them blocking other elements
    // from receiving events.
    constructor(cursor: HTMLElement, cursorRing: HTMLElement, animationDuration = 0.2, ringSizeMultiplier = 2) {
        super(cursor);
        this.dotCursorElement = cursor;
        this.cursorRing = cursorRing;
        this.ringSizeMultiplier = ringSizeMultiplier;
        this.cursorStartSize = this.GetDimensions(this.dotCursorElement);

        this.animationSpeed[0] = this.cursorStartSize[0] / 2 / (animationDuration * 30);
        this.animationSpeed[1] = this.cursorStartSize[1] / 2 / (animationDuration * 30);

        TouchFree.RegisterEventCallback('HandFound', this.ShowCursor.bind(this));
        TouchFree.RegisterEventCallback('HandsLost', this.HideCursor.bind(this));
    }

    // Function: UpdateCursor
    // Used to update the cursor when recieving a "MOVE" <ClientInputAction>. Updates the
    // cursor's position, as well as the size of the ring based on the current ProgressToClick.
    protected UpdateCursor(inputAction: TouchFreeInputAction): void {
        if (!this.enabled) return;
        //progressToClick is between 0 and 1. Click triggered at progressToClick = 1
        const ringScaler = MapRangeToRange(inputAction.ProgressToClick, 0, 1, this.ringSizeMultiplier, 1);

        this.cursorRing.style.opacity = inputAction.ProgressToClick.toString();

        const [cursorWidth, cursorHeight] = this.GetDimensions(this.dotCursorElement);
        const [cursorRingWidth, cursorRingHeight] = this.GetDimensions(this.cursorRing);

        this.cursorRing.style.width = cursorWidth * ringScaler + 'px';
        this.cursorRing.style.height = cursorHeight * ringScaler + 'px';

        this.cursorRing.style.left = inputAction.CursorPosition[0] - cursorRingWidth / 2 + 'px';
        this.cursorRing.style.top = inputAction.CursorPosition[1] - cursorRingHeight / 2 + 'px';

        super.UpdateCursor(inputAction);
    }

    // Function: HandleInputAction
    // This override replaces the basic functionality of the <TouchlessCursor>, making the
    // cursor's ring scale dynamically with the current ProgressToClick and creating a
    // "shrink" animation when a "DOWN" event is received, and a "grow" animation when an "UP"
    // is received.
    //
    // When a "CANCEL" event is received, the cursor is hidden as it suggests the hand has been lost.
    // When any other event is received and the cursor is hidden, the cursor is shown again.
    protected HandleInputAction(inputData: TouchFreeInputAction): void {
        switch (inputData.InputType) {
            case InputType.MOVE:
                this.UpdateCursor(inputData);
                break;
            case InputType.DOWN:
                this.SetCursorSize(0, 0, this.cursorRing);

                if (this.currentAnimationInterval !== -1) {
                    clearInterval(this.currentAnimationInterval);
                }

                this.currentAnimationInterval = setInterval(
                    this.ShrinkCursor.bind(this) as TimerHandler,
                    this.animationUpdateDuration
                );
                break;
            case InputType.UP:
                if (this.currentAnimationInterval !== -1) {
                    this.growQueued = true;
                } else {
                    this.growQueued = false;
                    this.currentAnimationInterval = setInterval(
                        this.GrowCursor.bind(this) as TimerHandler,
                        this.animationUpdateDuration
                    );
                }
                break;

            case InputType.CANCEL:
                break;
        }
    }

    // Function: ShrinkCursor
    // Shrinks the cursor to half of its original size.
    // This is performed over a duration set in the <constructor>.
    ShrinkCursor(): void {
        if (!this.enabled) return;
        let [newWidth, newHeight] = this.GetDimensions(this.dotCursorElement);

        if (newWidth > this.cursorStartSize[0] / 2) {
            newWidth -= this.animationSpeed[0];
        }

        if (newHeight > this.cursorStartSize[1] / 2) {
            newHeight -= this.animationSpeed[1];
        }

        this.SetCursorSize(newWidth, newHeight, this.dotCursorElement);

        if (newWidth <= this.cursorStartSize[0] / 2 && newHeight <= this.cursorStartSize[1] / 2) {
            clearInterval(this.currentAnimationInterval);

            newWidth = this.cursorStartSize[0] / 2;
            newHeight = this.cursorStartSize[1] / 2;

            this.SetCursorSize(newWidth, newHeight, this.dotCursorElement);

            if (this.growQueued) {
                this.growQueued = false;
                this.currentAnimationInterval = setInterval(
                    this.GrowCursor.bind(this) as TimerHandler,
                    this.animationUpdateDuration
                );
            } else {
                this.currentAnimationInterval = -1;
            }
        }
    }

    // Function: GrowCursor
    // Grows the cursor to its original size over time set via the <constructor>.
    GrowCursor(): void {
        if (!this.enabled) return;
        let [newWidth, newHeight] = this.GetDimensions(this.dotCursorElement);

        if (newWidth < this.cursorStartSize[0]) {
            newWidth += this.animationSpeed[0];
        }

        if (newHeight < this.cursorStartSize[1]) {
            newHeight += this.animationSpeed[1];
        }

        this.SetCursorSize(newWidth, newHeight, this.dotCursorElement);

        if (newWidth >= this.cursorStartSize[0] && newHeight >= this.cursorStartSize[1]) {
            clearInterval(this.currentAnimationInterval);

            this.SetCursorSize(this.cursorStartSize[0], this.cursorStartSize[1], this.dotCursorElement);

            this.currentAnimationInterval = -1;
            this.growQueued = false;
        }
    }

    private SetCursorSize(newWidth: number, newHeight: number, cursorToChange: HTMLElement): void {
        const deltaX = Math.round((parseFloat(cursorToChange.style.width) - newWidth) * 5) / 10;
        const deltaY = Math.round((parseFloat(cursorToChange.style.height) - newHeight) * 5) / 10;
        const cursorPosX = cursorToChange.offsetLeft + deltaX;
        const cursorPosY = cursorToChange.offsetTop + deltaY;

        cursorToChange.style.width = newWidth + 'px';
        cursorToChange.style.left = cursorPosX + 'px';

        cursorToChange.style.height = newHeight + 'px';
        cursorToChange.style.top = cursorPosY + 'px';
    }

    // Function: ShowCursor
    // Used to make the cursor visible, fades over time
    ShowCursor(): void {
        this.shouldShow = true;
        if (!this.enabled) return;
        clearInterval(this.currentFadingInterval);

        this.currentFadingInterval = setInterval(
            this.FadeCursorIn.bind(this) as TimerHandler,
            this.animationUpdateDuration
        );
    }

    // Function: HideCursor
    // Used to make the cursor invisible, fades over time
    HideCursor(): void {
        this.shouldShow = false;
        if (parseFloat(this.dotCursorElement.style.opacity) !== 0) {
            clearInterval(this.currentFadingInterval);

            this.currentFadingInterval = setInterval(
                this.FadeCursorOut.bind(this) as TimerHandler,
                this.animationUpdateDuration
            );
        }
    }

    private FadeCursorIn(): void {
        let currentOpacity = parseFloat(this.dotCursorElement.style.opacity);
        currentOpacity = currentOpacity ? currentOpacity : 0;
        currentOpacity += 0.05;

        this.dotCursorElement.style.opacity = currentOpacity.toString();

        if (currentOpacity >= 1) {
            clearInterval(this.currentFadingInterval);
            this.dotCursorElement.style.opacity = '1.0';
            this.currentFadingInterval = -1;
        }
    }

    private FadeCursorOut(): void {
        let currentOpacity = parseFloat(this.dotCursorElement.style.opacity);
        currentOpacity = currentOpacity ? currentOpacity : 1;
        currentOpacity -= 0.05;

        this.dotCursorElement.style.opacity = currentOpacity.toString();

        if (parseFloat(this.cursorRing.style.opacity) > 0) {
            this.cursorRing.style.opacity = currentOpacity.toString();
        }

        if (currentOpacity <= 0) {
            clearInterval(this.currentFadingInterval);
            this.dotCursorElement.style.opacity = '0.0';
            this.currentFadingInterval = -1;
        }
    }
}
